### PR TITLE
Support No reference Types in RUSTFLAGS by going from wasm->wat->wasm

### PR DESCRIPTION
**Pull request by rauljordan**
## Description

This PR honors wasms that are built with `export RUSTFLAGS=-Ctarget-feature=-reference-types` by removing reference type mentions through a conversion from wasm->wat->wasm. This is accompanied by also locking down the Rust toolchain of cargo stylus to 1.83 and a PR  to update stylus-hello-world to 1.83 with the specific Rust flag to disable reference types. This means that at the moment, with this change, there is no maximum Rust version limiting Stylus contracts or the SDK.

Huge thanks to @PlasmaPower for the suggestion.

### CODE REVIEW

**Comment by rauljordan** [comment]
Ready again @rory-ocl and @joshuacolvin0

File: main/src/project.rs [hunk 1]
```diff
@@ -101,19 +105,24 @@ pub fn build_dylib(cfg: BuildConfig) -> Result<PathBuf> {
...
     // Gets the files in the release folder.
     let release_files: Vec<PathBuf> = fs::read_dir(&release_path)
-        .map_err(|e| eyre!("could not read deps dir: {e}"))?
+        .map_err(|e| eyre!("could not read release deps dir: {e}"))?
         .filter_map(|r| r.ok())
         .map(|r| r.path())
         .filter(|r| r.is_file())
         .collect();
 
+    println!("Checking project name {}", project_name);
```
Comment (rauljordan) [inline main/src/project.rs:114]:
```suggestion
```

File: main/src/project.rs [hunk 2]
```diff
@@ -326,6 +350,11 @@ pub fn compress_wasm(wasm: &PathBuf, project_hash: [u8; 32]) -> Result<(Vec<u8>,
     let wasm =
         fs::read(wasm).wrap_err_with(|| eyre!("failed to read Wasm {}", wasm.to_string_lossy()))?;
 
+    let wat_str =
+        wasmprinter::print_bytes(&wasm).map_err(|e| eyre!("failed to convert Wasm to Wat: {e}"))?;
+    let wasm = wasmer::wat2wasm(wat_str.as_bytes())
+        .map_err(|e| eyre!("failed to convert Wat to Wasm: {e}"))?;
```
Comment (gligneul) [inline main/src/project.rs:356]: We should add a comment explaining this because it isn't intuitive.

**Review by gligneul — CHANGES_REQUESTED** [review]

File: main/src/check.rs [hunk 3]
```diff
@@ -62,12 +62,6 @@ pub async fn check(cfg: &CheckConfig) -> Result<ContractCheck> {
         greyln!("reading wasm file at {}", wasm.to_string_lossy().lavender());
     }
 
-    // Next, we include the project's hash as a custom section
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply main/src/check.rs:65]: Moved it further down, into the compress_wasm function where it makes more sense

**Review by gligneul — APPROVED** [review]
LGTM

**Pull request merged**